src/en/authors-charm-composing.md

```diff
@@ -0,0 +1,262 @@
+
+The `apache-php` layer sets an `apache.available` state when your application
+is done being installed, and the `mysql` interface sets a similar
+`{relation_name}.available` state (where `{relation_name}` depends on what you
+called the relation in your `metadata.yaml`) to indicate that MySQL has
+provided you with a complete set of database credentials.
+
+Thus, by watching for both of those states, you can easily tell when the right
+time to set up your application is:
+
+```python
+@when('apache.available', 'database.available')
+def setup_vanilla(mysql):
+    pass
+```
+
+You'll notice that the MySQL relation is passed in so that you can easily access
+the database connection information.
+
+Since the Apache installation and MySQL relation negotiation are all handled for
+us, the only thing left to do is create the `conf/config.php` file for Vanilla.
+To make things easy, we will render a template to populate the values.  We'll
+also handle a few more state changes to provide better status reporting and
+react to our database going away in an intelligent manner:
+
+```python
+import pwd
+import os
+from charmhelpers.core.hookenv import status_set
+from charmhelpers.core.templating import render
+from charms.reactive import when, when_not
+from charms.reactive import set_state, remove_state
+
+
+@when('apache.available', 'database.available')
+def setup_vanilla(mysql):
+    render(source='vanilla_config.php',
+           target='/var/www/vanilla/conf/config.php',
+           owner='www-data',
+           perms=0o775,
+           context={
+               'db': mysql,
+           })
+    uid = pwd.getpwnam('www-data').pw_uid
+    os.chown('/var/www/vanilla/cache', uid, -1)
+    os.chown('/var/www/vanilla/uploads', uid, -1)
+    set_state('apache.start')
+    status_set('maintenance', 'Starting Apache')
+
+
+@when('apache.available')
+@when_not('database.connected')
+def missing_mysql():
+    remove_state('apache.start')
+    status_set('blocked', 'Please add relation to MySQL')
+
+
+@when('database.connected')
+@when_not('database.available')
+def waiting_mysql(mysql):
+    remove_state('apache.start')
+    status_set('waiting', 'Waiting for MySQL')
+
+
+@when('apache.started')
+def started():
+    status_set('active', 'Ready')
+```
+
+By setting the `apache.start` state, we are letting the `apache-php` layer know
+that we have finished configuring the application and it is ready to start.
+
+The `templates/vanilla_config.php` file is straightforward:
+
+```php
+<?php if (!defined('APPLICATION')) exit();
+$Configuration['Database']['Host'] = '{{ db.host() }}';
+$Configuration['Database']['Name'] = '{{ db.database() }}';
+$Configuration['Database']['User'] = '{{ db.user() }}';
+$Configuration['Database']['Password'] = '{{ db.password() }}';
+?>
+```
+
+And that's all you need to get Vanilla up and running using layers!
+The final directory structure looks like this:
+
+![directory tree](./media/author-charm-composing-01.png)
```

Comment:
I'd argue having the file tree as actual text wrapped in a code block instead of a picture. It'll make it easier to update overtime, but that's just an opinion of mine.